chips/lowrisc/src/hmac.rs

```diff
@@ -0,0 +1,252 @@
+        FIFO_FULL OFFSET(1) NUMBITS(1) [],
+        FIFO_DEPTH OFFSET(4) NUMBITS(5) []
+    ]
+];
+
+pub struct Hmac<'a> {
+    registers: StaticRef<HmacRegisters>,
+
+    client: OptionalCell<&'a dyn hil::digest::Client<'a, [u8; 32]>>,
+
+    data: Cell<Option<LeasableBuffer<'static, u8>>>,
+    data_len: Cell<usize>,
+    data_index: Cell<usize>,
+
+    digest: Cell<Option<&'static mut [u8; 32]>>,
+}
+
+impl Hmac<'a> {
+    pub const fn new(base: StaticRef<HmacRegisters>) -> Hmac<'a> {
+        Hmac {
+            registers: base,
+            client: OptionalCell::empty(),
+            data: Cell::new(None),
+            data_len: Cell::new(0),
+            data_index: Cell::new(0),
+            digest: Cell::new(None),
+        }
+    }
+
+    fn data_progress(&self) {
+        let regs = self.registers;
+        let idx = self.data_index.get();
+        let len = self.data_len.get();
+
+        let slice = self.data.take().unwrap().take();
+
+        if idx < len {
+            let data_len = len - idx;
+
+            for i in 0..(data_len / 4) {
+                if regs.status.is_set(STATUS::FIFO_FULL) {
+                    // Due to: https://github.com/lowRISC/opentitan/issues/1276
+                    //   we can't get back to processing the data as there is no
+                    //   FIFO not full interrupt.
+                    // Let's just keep going and put up with the back pressure.
+                    // break;
+                }
+
+                let data_idx = i * 4;
+
+                let mut d = (slice[data_idx + 0] as u32) << 0;
+                d |= (slice[data_idx + 1] as u32) << 8;
+                d |= (slice[data_idx + 2] as u32) << 16;
+                d |= (slice[data_idx + 3] as u32) << 24;
+
+                regs.msg_fifo.set(d);
+                self.data_index.set(data_idx + 4)
+            }
+
+            let idx = self.data_index.get();
+
+            for i in 0..(data_len % 4) {
+                let data_idx = idx + i;
+                let d = (slice[data_idx]) as u32;
+
+                regs.msg_fifo.set(d);
+                self.data_index.set(data_idx + 1)
+            }
+        }
+
+        self.client.map(move |client| {
+            client.add_data_done(Ok(()), slice);
+        });
+    }
+
+    pub fn handle_interrupt(&self) {
+        let regs = self.registers;
+        let intrs = regs.intr_state.extract();
+
+        regs.intr_enable
+            .modify(INTR_ENABLE::HMAC_DONE::CLEAR + INTR_ENABLE::HMAC_ERR::CLEAR);
+
+        if intrs.is_set(INTR_STATE::HMAC_DONE) {
+            self.client.map(|client| {
+                let digest = self.digest.take().unwrap();
+
+                for i in 0..8 {
+                    let d = regs.digest[i].get().to_ne_bytes();
+
+                    let idx = i * 4;
+
+                    digest[idx + 0] = d[3];
+                    digest[idx + 1] = d[2];
+                    digest[idx + 2] = d[1];
+                    digest[idx + 3] = d[0];
```

Comment:
This reverts the byte order of `d`. Couldn't this block be replaced by:

```rust
let d = regs.digest[i].get().swap_bytes().to_ne_bytes();

let idx = i * 4;

digest[idx + 0] = d[0];
digest[idx + 1] = d[1];
digest[idx + 2] = d[2];
digest[idx + 3] = d[3];
```

Or (with whatever endianness is not native):

```rust
let d = regs.digest[i].get().to_(be|le)_bytes();
```

Reply:
Or maybe all this endianness swapping can be avoided by clearing the `CFG::DIGEST_SWAP` bit instead of setting it?
I'm not familiar with OT and unfortunately the FPGA boards I have are not compatible to try it.

Reply:
I have removed the endlessness swap